cpp/unreal_plugins/UrdfBot/Source/UrdfParser.h

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <Eigen/Dense>
+#include <map>
+#include <vector>
+
+#include <CoreMinimal.h>
+
+class FXmlNode;
+
+struct UrdfJointDesc;
+struct UrdfLinkDesc;
+struct UrdfMaterialDesc;
+
+enum UrdfGeometryType
+{
+    INVALID_GEOMETRY = 0,
+    BOX,
+    SPHERE,
+    CYLINDER,
+    MESH
+};
+
+enum UrdfJointType
+{
+    // Some of these (such as FIXED) are already macros
+    INVALID_TYPE = 0,
+    REVOLUTE_TYPE,
+    CONTINUOUS_TYPE,
+    PRISMATIC_TYPE,
+    FIXED_TYPE,
+    FLOATING_TYPE,
+    PLANAR_TYPE
+};
+
+struct UrdfGeometry
+{
+    UrdfGeometryType type_;
+    FVector size_; // requried by BOX
+    float radius_; // requried by SPHERE, CYLINDER
+    float length_; // requried by CYLINDER
+    std::string file_name_; // requried by MESH
+};
+
+struct UrdfLinkDesc
+{
+    std::string name_;
+
+    // non-derived data (inertial, visual, etc)...
+    FTransform origin_ = FTransform::Identity;
+    float mass_;
+    Eigen::Matrix3f inertia_;
+    FTransform visual_origin_ = FTransform::Identity;
+    UrdfGeometry visual_geometry_;
+    FTransform collision_origin_ = FTransform::Identity;
+    UrdfGeometry collision_geometry_;
+    std::string material_name_;
+
+    // derived data
+    UrdfLinkDesc* parent_link_desc_ = nullptr;
+    UrdfJointDesc* parent_joint_desc_ = nullptr;
+
+    std::vector<UrdfLinkDesc*> child_link_descs_ = {};
+    std::vector<UrdfJointDesc*> child_joint_descs_ = {};
+
+    UrdfMaterialDesc* material_desc_ = nullptr;
```

Comment:
`material_desc_` should be included in the derived data for `UrdfVisualDesc`, not `UrdfLinkDesc`

(I think I got this wrong in my initial sketch)